### PR TITLE
add sharded optim v3

### PR DESCRIPTION
**Pull request by ver217**
Use `ShardedTensor` API. Support `ShardedModel` and normal model.

### CODE REVIEW

File: colossalai/zero/sharded_optim/sharded_optim_v3.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                 min_scale: float = 1,
+                 growth_factor: float = 2,
+                 backoff_factor: float = 0.5,
+                 growth_interval: float = 1000,
+                 hysteresis: float = 2,
+                 max_scale: int = 2**32,
+                 dp_process_group: Optional[ProcessGroup] = None,
+                 mp_process_group: Optional[ProcessGroup] = None) -> None:
+        super().__init__(optimizer)
+        self.model: Union[nn.Module, ShardedModelV2] = sharded_model
```
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v3.py:43]: I just wonder if we can do it in this way.
If the class receive a Module type sharded_model. It just converts it into a ShardedModelV2

sharded_model = ShardedModelV2(sharded_model， ..)

File: colossalai/zero/sharded_optim/sharded_optim_v3.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                    p.data = self.master_params[p]
+                else:
+                    self.shard_strategy.shard([self.master_params[p]])
+                    p.data = self.master_params[p].payload
+        ret = self.optim.step(*args, **kwargs)
+        # Write master param to payload
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                if hasattr(p, 'ca_attr'):
+                    p.ca_attr.data = p.data
```
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v3.py:99]: Did you make sure the param copy is fp32 -> fp16?

Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v3.py:99]: If it is ZeRO3, it should copy cpu fp32 -> gpu fp16.

Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v3.py:99]: the setter of ShardedParam data is not good, you can add a new function as copy_data(self, t)?

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v3.py:99]: OK

File: colossalai/zero/sharded_optim/sharded_optim_v3.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+from ..sharded_param import ShardedTensor
+from ._utils import has_inf_or_nan
+
+
+class OptimState(Enum):
+    SCALED = 1
+    UNSCALED = 2
+
+
+class ShardedOptimizerV2(ColossalaiOptimizer):
```
Comment (FrankLeeeee) [inline colossalai/zero/sharded_optim/sharded_optim_v3.py:27]: File name is V3 but the class name is V2.

**Pull request closed without merging**